src/BenchmarkDotNet/Disassemblers/ClrMdV3Disassembler.cs

```diff
@@ -213,60 +215,30 @@ private IEnumerable<Asm> Decode(ILToNativeMap map, State state, int depth, ClrMe
+            // it's simply easier to get next instruction when decoding ;)
+
+            var hotColdInfo = method.HotColdInfo;
+            if (hotColdInfo.HotSize > 0 && hotColdInfo.HotStart > 0)
             {
-                startAddress = method.NativeCode;
-                endAddress = ulong.MaxValue;
+                return hotColdInfo.ColdSize <= 0
+                    ? new[] { new ILToNativeMap() { StartAddress = hotColdInfo.HotStart, EndAddress = hotColdInfo.HotStart + hotColdInfo.HotSize, ILOffset = -1 } }
+                    : new[]
+                      {
+                            new ILToNativeMap() { StartAddress = hotColdInfo.HotStart, EndAddress = hotColdInfo.HotStart + hotColdInfo.HotSize, ILOffset = -1 },
+                            new ILToNativeMap() { StartAddress = hotColdInfo.ColdStart, EndAddress = hotColdInfo.ColdStart + hotColdInfo.ColdSize, ILOffset = -1 }
+                      };
             }
 
-            ILToNativeMap[] sortedMaps = method.ILOffsetMap // CanBeDisassembled ensures that there is at least one map in ILOffsetMap
-                .Where(map => map.StartAddress >= startAddress && map.StartAddress < endAddress) // can be false for Tier 0 maps, EndAddress is not checked on purpose here
-                .Where(map => map.StartAddress < map.EndAddress) // some maps have 0 length (they don't have corresponding assembly code?)
+            return method.ILOffsetMap
+                .Where(map => map.StartAddress < map.EndAddress) // some maps have 0 length?
                 .OrderBy(map => map.StartAddress) // we need to print in the machine code order, not IL! #536
-                .Select(map => new ILToNativeMap()
-                {
-                    StartAddress = map.StartAddress,
-                    // some maps have EndAddress > codeHeaderData.MethodStart + codeHeaderData.MethodSize and contain garbage (#2074). They need to be fixed!
-                    EndAddress = Math.Min(map.EndAddress, endAddress),
-                    ILOffset = map.ILOffset
-                })
                 .ToArray();
-
-            if (sortedMaps.Length == 0)
-            {
-                // In such situation ILOffsetMap most likely describes Tier 0, while CodeHeaderData Tier 1.
-                // Since we care about Tier 1 (if it's present), we "fake" a Tier 1 map.
-                return new[] { new ILToNativeMap() { StartAddress = startAddress, EndAddress = endAddress } };
-            }
-            else if (sortedMaps[0].StartAddress != startAddress || (sortedMaps[sortedMaps.Length - 1].EndAddress != endAddress && endAddress != ulong.MaxValue))
-            {
-                // In such situation ILOffsetMap most likely is missing few bytes. We just "extend" it to avoid producing "bad" instructions.
-                return new[] { new ILToNativeMap() { StartAddress = startAddress, EndAddress = endAddress } };
-            }
-
-            return sortedMaps;
```

Comment:
It took me a lot of time to workaround all ClrMd bugs. If we really want to remove these workarounds we need to do a very extensive testing: 

1. Run a simple benchmark and ask BDN to disassemble ALL methods for main. It should produce a markdown file with all CLR methods disassembled.
2. Same as above, but for this PR.
3. Diff two produced .md files and ensure that all instructions are correctly resolved and there are no missing things.